Backport FW filter package changes: allow arrays process by clean()

photodude opened the pull request:
Updates the CMS Framework package to [framework/filter 1.2.0](https://github.com/joomla-framework/filter/commit/45a504f0cd6fa2312a0260955e957435ca2cedc9) with Composer

Backports Framework changes: Allow arrays of numbers to be processed by clean() and associated tests.
joomla-framework/filter#13
Backports additional changes from [framework/filter 1.2.0](https://github.com/joomla-framework/filter/commit/45a504f0cd6fa2312a0260955e957435ca2cedc9)
- Add array support to most filters
  - Fixes Array to String conversion notices in clean() for most cases
- Updates pcre patterns for numbers
  - Fixes floats case to properly handle exponent type floats rather than just decimal floats
- Adds new general cases to test and verify the array filter functionality
- Code style: fixes missing comma at end of array lists.
  *Fixes PHP7 issues

Testing
https://docs.joomla.org/Testing_Joomla!_patches
Do a general check that the various existing forms in the CMS function as expected.

photodude committed:
Backport FW change: num arrays process by clean()

Backports Framework changes: Allow arrays of numbers to be processed by clean() 
https://github.com/joomla-framework/filter/pull/13

* Adds basic first level processing of array's of numbers
  * Fixes Array to String conversion notices in clean() for number cases
* Updates pcre patterns for numbers
  * Fixes floats case to properly handle exponent type floats rather than just decimal floats
* Adds new general cases to test and verify the array of numbers functionality
* Code style: fixes missing comma at end of array lists.

photodude committed:
Backport FW change: num arrays process by clean()

Adds new and missing tests from the framework. https://github.com/joomla-framework/filter/pull/13

photodude commented:
@mbabker Here is the backport of the recent framework changes for the input filter clean() method and test.

photodude commented:
@RickR2H would you be willing to test this?

RickR2H commented:
@photodude I'll test this if I can. What is the best procedure to test as I don't have in depth knowledge of the code as I'm only a frontend developer ;)

photodude commented:
I would just do a general check that the various existing forms in the CMS function as expected. 
I'm not sure there are any easy tests beyond that which could have a "clean" method triggered for testing.

photodude committed:
Update the return type descriptions

photodude committed:
Add array support to most filters

Backports additional changes from Framework/filter 1.2.0
Fix the line breaks in checkAttribute() so it's properly on the boolean

photodude committed:
Add array support to most filter tests

Backports changes from framework/filter 1.2.0

photodude committed:
Need to use Joomla\String\StringHelper;

Replaces deprecated JString with StringHelper

photodude committed:
Fix Whitespace found at end of line 571

photodude commented:
@mbabker @wilsonge Updated to match changes in framework/filter 1.2.0

photodude commented:
@brianteeman Unlike @wilsonge 's PR #8734 for updating to the current the framework package. This filter class is used in the CMS core.
This should have consideration for merging for 3.5

brianteeman commented:
Not sure why that comment is addressed at me. I have no say on what gets merged or when

photodude commented:
I only Pinged you since You had asked in #8734 if that PR "need to be merged before 3.5"

wilsonge committed:
Update Joomla Filter package to 1.2 with PHP 7 Fixes

wilsonge commented:
Merge this: https://github.com/photodude/joomla-cms/pull/1

Get some testers in here. I've gotta focus on the sql upgrade. But added a 3.5 blocker tag as even though the FW filter package isn't being used by core - let's not run into PHP 7 issues because 3rd parties are using the framework filter package :)

photodude commented:
@RickR2H would you still be willing to test this?

Stevec4 commented:
I received an error after applying this patch.
`Fatal error: Uncaught Error: Class 'ComposerAutoloaderInit205c915b9c7d3e718e7c95793ee67ffe' not found in C:\xampp\htdocs\sites\aviondemo\libraries\vendor\autoload.php:7 Stack trace: #0 C:\xampp\htdocs\sites\aviondemo\libraries\cms.php(33): require() #1 C:\xampp\htdocs\sites\aviondemo\administrator\includes\framework.php(23): require_once('C:\\xampp\\htdocs...') #2 C:\xampp\htdocs\sites\aviondemo\administrator\index.php(40): require_once('C:\\xampp\\htdocs...') #3 {main} thrown in C:\xampp\htdocs\sites\aviondemo\libraries\vendor\autoload.php on line 7
`
I applied the patch to a 3.5 beta site running php 7.0.2
Let me know if I can provide any more info.
Steve

photodude commented:
@Stevec4 Thanks for testing, I didn't create the part of this PR for the composer changes So I'll have to dig into what's going on.

I'm not sure, but I think this patch might need to be applied against Stagging, before the composer changes were added it would have been a test against 3.4.8 or possibly 3.5B2 The composer changes and the PHP7 compatibility add some different dynamics.

I'll see if I can rebase this soon since there are some new merge conflicts.

photodude commented:
@Stevec4 I've rebased to staging and fixed the Merge conflicts. I think the issue you reported should be fixed. I tested against Staging, So I'm not sure about success for testing from J3.5B2 and then applying the patch.

Stevec4 commented:
@photodude I will re apply and advise. I can test it on a 348 install as well.
Steve

anibalsanchez commented:
I have tested this item :white_check_mark: successfully on 23ca41a717197d7c27d370e7be858373199630ff<br /><br />Test OK<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8681">issues.joomla.org/joomla-cms/8681</a>.</sub>

photodude commented:
@Stevec4 thanks for the test. Would you please go to the issue tracker and mark it as a successful test. 
https://issues.joomla.org/tracker/joomla-cms/8681

brianteeman commented:
Manually added the test from @Stevec4 

Marking RTC - thanks everyone<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8681">issues.joomla.org/joomla-cms/8681</a>.</sub>

wilsonge commented:
Merged -thanks all!

photodude commented:
Thanks Everyone. I appreciate all the work and testing.

beat commented on libraries/joomla/filter/input.php:
```diff
@@ -103,7 +105,7 @@ class JFilterInput
 		'style',
 		'title',
-		'xml'
+		'xml',
```
Why did a comma got added here  and just below ?

wilsonge replied:
I personally prefer it when you have multiple array elements. I don't think our code style policy defines either way...

photodude replied:
It's a [Joomla Coding styles](http://joomla.github.io/coding-standards/?coding-standards/chapters/php.md) option.
"When splitting array definitions onto several lines, the last value may also have a trailing comma."
I follow that the last value _**should**_ have a trailing comma

beat commented on libraries/joomla/filter/input.php:
```diff
@@ -234,57 +236,214 @@ public function clean($source, $type = 'string')
+					$result = array();
+
+					// Itterate through the array
+					foreach ($source as $eachString)
```
Shouldn't this keep the keys (and cleaning the keys) ?

wilsonge replied:
No. We play safe rather than sorry. We don't know what is in the keys - it could be a integer or a string (and we don't know how strict to filter the key). Play safe. Start a fresh array with the sanitised integer strings. We have no b/c constraints here - let's not try and make things too complex

photodude replied:
We are cleaning the array values not the array keys. It's assumed that the keys are not important the context of input filtering of input values. 

Personally, I think that this is a start, but I would love to see the whole [Joomla! Framework Filter Package](https://github.com/joomla-framework/filter) totally rewritten with recursive support for arrays that have mixed types. Just not sure I want to try tackling that rewrite.

But that recursive support for arrays is truly a discussion and a PR for the [Joomla! Framework Filter Package](https://github.com/joomla-framework/filter).

The JFilterInput in the CMS really needs to be rewritten so that it's extending the Framework filter package with CMS specific additions or changes rather than duplicating them modifying the code from the Framework. Just not sure I want to try tackling that rewrite either.

wilsonge replied:
Extending the frameworks package isn't the worst - it's just the underscore function names that you can't actually rewrite b/c. or at least you can but you end up at the point where you may as well have not extended the class to start with....

photodude replied:
@wilsonge I gave it a [quick shot for an extends rewrite](https://github.com/photodude/joomla-cms/blob/patch-2/libraries/joomla/filter/input.php), But I'm sure I missed something since [it's failing the Travis unit tests](https://travis-ci.org/photodude/joomla-cms/jobs/112340941)

photodude replied:
@wilsonge If you have the time and could give me some guidance on what I missed in changing this to [Extending the frameworks package](https://github.com/photodude/joomla-cms/tree/patch-2/libraries/joomla/filter) I would appreciate it.

wilsonge replied:
Can you ping me tomorrow to remind me. I'm trying to finish off the utf8mb4 stuff right now and there's a decent chance it will slip my mind

photodude replied:
Will do, Good luck with the utf8mb4 stuff.

photodude replied:
@wilsonge Ping as requested.

wilsonge replied:
OK So even without the unit test failures - I can already see a big problem (which is why I gave up) - you're removing all the methods starting with underscores. Sadly the project at this moment with the current b/c rules considers that to be a b/c break :(

photodude replied:
Are you sure you're looking at the right code? All of the methods starting with underscores are still there, I left those in since they are still listed as @deprecated 4.0.

[Example line 454 from the input filter](https://github.com/photodude/joomla-cms/blob/patch-2/libraries/joomla/filter/input.php#L454) that shows the methods starting with underscores

mbabker replied:
They aren't called though.  So in the off chance someone has extended one of the classes and expects the underscored methods to be executed they wouldn't be.  We've got proxies all over the place because of that (`JRouter` and `JUri` are two examples).

photodude replied:
This is my first time extending in PHP (outside of JTable for a component) or from a Framework  class. So maybe I'm not following what you mean by "They aren't called though"

Are you saying that I need to change the internal method for those methods starting with underscores? and as such in its current state they are no longer aliased to the correct methods?

or are you saying that I need to also keep the related methods not starting with underscores and alias those to the framework?

mbabker replied:
It's a B/C break to go from using `_remove()` to `remove()` without some sort of proxy.  Generally that's been to deprecate the underscore method and just have that proxy to the non-underscored method.

Even though you haven't removed the old methods, by not calling them, you introduce a change in the internal behaviors.  In the off chance someone's extended `JFilterInput`, they expect `_remove()` to be called at certain places.  With that patch merged as is, their custom method would no longer be called and they'd have to update their internal references to the new name.  Go back to the examples I pointed out in `JUri`.  We've made that extend the Framework's Uri class but had to put some internal proxies in place to not break API compatibility (including keeping a method that the Framework changed from public to protected exposed as a public method and proxies to handle with calling the underscored methods).

It's one thing to keep the old methods in place, but it's another to refactor in a way that it changes the API behavior chain.  `JFilterInput` still needs to call `_remove()` until 4.0, and unfortunately that's going to mean keeping a lot of duplicated code compared to the parent Framework class.

photodude replied:
I'm not sure I fully understand, but The general concept that I think you're telling me is 

For something like the existing code
```
	protected function _cleanTags($source)
	{
		return $this->cleanTags($source);
	}
```
I need to change to something like
```
	protected function _cleanTags($source)
	{
		return parent::cleanTags($source);
	}
```
[As was done in `JUri`](https://github.com/photodude/joomla-cms/blob/patch-2/libraries/joomla/uri/uri.php#L329)

And for the methods that use the methods starting with underscores in their code I need to keep those methods in place.

Here is the list I have of the methods starting with underscores from the source code and where they are in the extended code
Input
- [_remove](https://github.com/photodude/joomla-cms/blob/patch-2/libraries/joomla/filter/input.php#L454)
- [_cleanTags](https://github.com/photodude/joomla-cms/blob/patch-2/libraries/joomla/filter/input.php#L469)
- [_cleanAttributes](https://github.com/photodude/joomla-cms/blob/patch-2/libraries/joomla/filter/input.php#L484)
- [_decode](https://github.com/photodude/joomla-cms/blob/patch-2/libraries/joomla/filter/input.php#L499)
- [_escapeAttributeValues](https://github.com/photodude/joomla-cms/blob/patch-2/libraries/joomla/filter/input.php#L557)
- [_stripCSSExpressions](https://github.com/photodude/joomla-cms/blob/patch-2/libraries/joomla/filter/input.php#L572)

output
- [_ampReplaceCallback](https://github.com/photodude/joomla-cms/blob/patch-2/libraries/joomla/filter/output.php#L88)

So those methods need the modification listed above.

The following methods should also be maintained due to the internal use of underscore methods (at least until they are listed as deprecated i.e. 4.0)
input
- remove()
- cleanTags()
- escapeAttributeValues()
output
- linkXHTMLSafe()

mbabker replied:
Leave it as `$this->cleanTags()`.  Calling `parent::cleanTags()` explicitly calls up the parent method and breaks OOP principles (`$this` will call it in the right place).  The one example you found in `JUri` should be patched with that change.

All the underscore methods should be deprecated and proxied to non-underscore.  I think that was done already anyway.  Places still calling `$this->_remove()` (for example) need to stay that way until 4.0.  So something like `JFilterInput::clean()` couldn't just use the Framework's `clean()` method without hacking that to call underscore methods, so the extended method has to stay in place (even if it's basically duplicated in full except for the underscore differences) too.  It's one of those tricky things to deal with.

photodude replied:
Thanks @mbabker for the guidance, 

I think all of the issues are fixed and all of the methods with things like `$this->_remove()` have been returned.

I did find that the `__construct` method had to fully overridden rather than calling `parent::__construct` with the additional parameters included
([The way I'm familiar with calling `__construct`](https://github.com/joomla/joomla-cms/blob/staging/components/com_banners/helpers/category.php#L26) )

I'll submit this as a PR
